SYCL/Regression/thread_safety_unchached_kernel_enqueue.cpp

```diff
@@ -0,0 +1,129 @@
+// REQUIRES: level_zero, level_zero_dev_kit
+
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %level_zero_options %s -o %t.out %threads_lib
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/online_compiler.hpp>
+
+#include <vector>
+
+using byte = unsigned char;
+
+const char *KernelSrc = "kernel void my_kernel( \
+    global double *a, global double *b, global double *c, \
+    int lda, int ldb, ldc, int k) { \
+    int i = get_global_id(0); \
+    int j = get_global_id(1); \
+    double sum = 0; \
+    for (int h = 0; h < k; h++) { \
+        sum += a[i * lda + h] * b[h * ldb + j]; \
+    } \
+    c[i * ldc + j] = sum;}";
+
+sycl::kernel getSYCLKernelWithIL(sycl::context &Context,
+                                 const std::vector<byte> &IL) {
+
+  ze_module_desc_t ZeModuleDesc = {};
+  ZeModuleDesc.format = ZE_MODULE_FORMAT_IL_SPIRV;
+  ZeModuleDesc.inputSize = IL.size();
+  ZeModuleDesc.pInputModule = IL.data();
+  ZeModuleDesc.pBuildFlags = "";
+  ZeModuleDesc.pConstants = nullptr;
+
+  assert(Context.get_devices().size() == 1 && "Expected to have only 1 device");
+  sycl::device Device = Context.get_devices()[0];
+  auto ZeDevice = Device.get_native<sycl::backend::level_zero>();
+  auto ZeContext = Context.get_native<sycl::backend::level_zero>();
+
+  ze_module_build_log_handle_t ZeBuildLog;
+  ze_module_handle_t ZeModule;
+  ze_result_t ZeResult = zeModuleCreate(ZeContext, ZeDevice, &ZeModuleDesc,
+                                        &ZeModule, &ZeBuildLog);
+  if (ZeResult != ZE_RESULT_SUCCESS)
+    throw sycl::runtime_error();
+  sycl::program SyclProgram =
+      sycl::level_zero::make<sycl::program>(Context, ZeModule);
+  return SyclProgram.get_kernel("my_kernel");
+}
+
+void runKernel(sycl::queue &Q, sycl::kernel Kernel, double *a, double *b,
+               double *c, int size_n) {
+  size_t n = size_n;
+  range local_group_size{1, 1};
+  range global_group_size{n, n};
+
+  constexpr int N = 4;
+  int InputArray[N] = {0, 1, 2, 3};
+  int OutputArray[N] = {};
+
+  sycl::buffer<int, 1> InputBuf(InputArray, sycl::range<1>(N));
+  sycl::buffer<int, 1> OutputBuf(OutputArray, sycl::range<1>(N));
+
+  Q.submit([&](sycl::handler &CGH) {
```

Comment:
Can we simplify the test a bit?:
1. Change the kernel to write values of it's arguments to some output buffer and then check "passed" arguments with what the kernel gets.
2. Instead of creating a program using Level Zero API, we could try to create a SYCL program then extract native handle from it then use this handle for creating an interop SYCL program.


BTW.
Wouldn't it be better to have such a test in the "unit-tests"? We could redefine setKernelArg and EnqueueKernel APIs to record kernel ID, thread ID, kernel argument ID and value and then check that the sequence is correct.